chore(build): maven 4 migration

sruehl opened the pull request:
https://maven.apache.org/guides/mini/guide-migration-to-mvn4.html

sruehl committed:
chore(build): maven 4 migration

https://maven.apache.org/guides/mini/guide-migration-to-mvn4.html

sruehl committed:
chore(build): maven 4 migration

https://maven.apache.org/guides/mini/guide-migration-to-mvn4.html

sruehl commented:
@chrisdutz @hutcheb seems that Maven4 breaks the way we do you code-generation integration tests. It seems that they can't access code-generation parent:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-invoker-plugin:3.9.1:install (integration-test) on project plc4x-code-generation-language-go: The following artifacts could not be resolved: org.apache.plc4x:plc4x-code-generation:pom: (absent): org.apache.plc4x:plc4x-code-generation:pom: was not found in https://repository.apache.org/content/repositories/releases during a previous attempt. This failure was cached in the local repository and resolution is not reattempted until the update interval of apache-releases has elapsed or updates are forced -> [Help 1]
```

sruehl committed:
chore(build): move deploy/install settings to the top

https://maven.apache.org/guides/mini/guide-migration-to-mvn4.html

sruehl committed:
chore(build): move deploy/install settings to the top

https://maven.apache.org/guides/mini/guide-migration-to-mvn4.html

sruehl committed:
chore(build): upgrade to newer rc

sruehl committed:
chore(build): upgrade to newer rc

hutcheb commented:
Just checking to see if this is an issue with the version of the maven plugin in the plc4x-build tools repo

sruehl committed:
chore(build): migrate poms added on develop since branch creation to maven 4

sruehl committed:
chore(build): apply mvnup inference optimizations

Applied 'mvnup apply' from Maven 4.0.0-rc-5:
- drop explicit modelVersion (inferred from the 4.1.0 POM namespace)
- drop subprojects lists that exactly match the direct child directories
  (automatic subproject discovery)

sruehl committed:
chore(build): drop obsolete installAtEnd/deployAtEnd workarounds, update README

The TODO workarounds forcing installAtEnd/deployAtEnd=false are no longer
needed: the effective maven-install/deploy-plugin 3.1.4 (pinned via
apache parent 39) defaults to false anyway, and a release-profile build
with -DinstallAtEnd=true (the Maven 4 plugin-line default) now completes
successfully on 4.0.0-rc-5, with installs correctly deferred to the end
of the reactor.

Also update the README build requirements to Maven 4.0.0-rc-5+.

sruehl committed:
fix(build): repair code-generation integration tests under Maven 4

Three distinct issues broke the CI platform workflows (all failing since
the branch was created):

1. maven-invoker-plugin:install resolved in-reactor dependency POMs
   through the workspace reader, which serves the raw 4.1.0 build POM.
   In repository-mode model building the omitted parent <version> is not
   inferred, producing an unresolvable versionless coordinate
   (org.apache.plc4x:plc4x-code-generation:pom:). Restore explicit
   parent versions across the code-generation subtree.

2. The invoker IT project POMs pointed <relativePath> four levels up,
   which from the cloned target/integration-tests location resolves to
   the wrong POM. Maven 3 silently fell back to repository resolution on
   the coordinate mismatch; Maven 4 follows the path and fails with a
   parent-cycle error. Use an empty <relativePath/> so the parent is
   resolved from the invoker-provisioned local repository.

3. The plc4py IT POM declared exec-maven-plugin twice, which Maven 4
   treats as a hard error. Merge into a single declaration.

sruehl committed:
fix(build): restore explicit parent versions in all reactor poms

Omitting the parent <version> (model 4.1.0 inference) breaks any plugin
that resolves in-reactor artifact descriptors through the repository
system: the workspace reader serves the raw build POM, and
repository-mode model building does not apply parent-version inference
(relativePath is not followed there). maven-invoker-plugin:install hits
this for every in-reactor dependency of the code-generation language
modules (first plc4x-code-generation, then plc4j once the full reactor
is in the session).

Keep the rest of the model 4.1.0 features (subprojects auto-discovery,
sources element, omitted modelVersion) which are unaffected.
